network-proxy-service/src/sshforwarding.rs

```diff
@@ -0,0 +1,150 @@
+    pub local_port: u16,
+}
+
+pub struct SshForwarding {
+    config: SshForwardingConfig,
+    ssh_client: Option<Handle<ClientHandler>>,
+    local_listener: Option<TcpListener>,
+}
+
+impl SshForwarding {
+    const DEFAULT_SSH_PORT: u16 = 22;
+
+    pub fn new(config: SshForwardingConfig) -> Self {
+        return Self { config: config, ssh_client: None, local_listener: None };
+    }
+
+
+    pub async fn prepare_ssh_client(&mut self) -> Result<(), Error> {
+        let ssh_addrs = Url::parse(&self.config.ssh_endpoint)?.socket_addrs(|| Some(Self::DEFAULT_SSH_PORT))?;
+        let ssh_addr = ssh_addrs.iter().next().ok_or(Error::InvalidSshEndpoint)?;
+        let config = Arc::new(client::Config::default());
+        let handler = ClientHandler {};
+        self.ssh_client = Some(client::connect( config, ssh_addr, handler).await?);
+
+        Ok(())
+    }
+
+    pub async fn prepare_local_listener(&mut self) -> Result<(), Error> {
+        let local_listen_addr: SocketAddr = format!("127.0.0.1:{}", self.config.local_port).parse()?;
+        self.local_listener = Some(TcpListener::bind(local_listen_addr).await?);
+
+        Ok(())
+    }
+
+    pub async fn authenticate(&mut self) -> Result<(), Error> {
+        let pem = decode(&self.config.ssh_private_key_base64)?;
+
+        let key_pair = Arc::new(key::KeyPair::RSA {
+            key: openssl::rsa::Rsa::private_key_from_pem(&pem)?,
+            hash: key::SignatureHash::SHA2_256,
+        });
+
+        let sc = self.ssh_client.as_mut().ok_or(Error::SshClientUnInitialized)?;
+        if !sc.authenticate_publickey(&self.config.ssh_user, key_pair).await? {
+            return Err(Error::InvalidSshCredential)
+        }
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl NetworkProxy for SshForwarding {
+    async fn prepare(&mut self) -> Result<(), Error> {
+        self.prepare_ssh_client().await?;
+        self.prepare_local_listener().await?;
+        self.authenticate().await?;
+        Ok(())
+    }
+
+    async fn start_serve(&mut self) -> Result<(), Error> {
+        let sc = self.ssh_client.as_mut().ok_or(Error::SshClientUnInitialized)?;
+        let ll = self.local_listener.as_mut().ok_or(Error::LocalListenerUnInitialized)?;
+        loop {
+            let (forward_stream, _) = ll.accept().await?;
+            let bastion_channel = sc.channel_open_direct_tcpip(
+                &self.config.remote_host,
+                self.config.remote_port as u32,
+                "127.0.0.1", 0).await?;
+            tokio::task::spawn(async move {
+                tunnel_streaming(forward_stream, bastion_channel).await.or_bail("tunnel_handle failed.");
+            });
+        }
+    }
+}
+
+async fn tunnel_streaming(mut forward_stream: TcpStream, mut bastion_channel: client::Channel) -> Result<(), Error>{
+    let mut buf_forward_stream = vec![0; 2048];
+
+    loop {
+        select! {
+            bytes_read = forward_stream.read(&mut buf_forward_stream).fuse() => match bytes_read? {
```

Comment:
This is calling the synchronous `read` function, not the async one. 
I think there's a simpler way of doing this, though. The `client::Channel` can be wrapped in another struct that implements `AsyncRead` and `AsyncWrite`. Once you have that, you should be able to use [`copy_bidirectional`](https://docs.rs/tokio/latest/tokio/io/fn.copy_bidirectional.html) to pass all the data around as a single task. LMK if you want to talk through that in more detail. 

I think you could also just change to using async read functions and keep the general structure you have here, if the `copy_bidirectional` approach is no good.